src/ray/gcs/gcs_server/gcs_autoscaler_state_manager.cc

```diff
@@ -0,0 +1,129 @@
+// Copyright 2023 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/gcs/gcs_server/gcs_autoscaler_state_manager.h"
+
+namespace ray {
+namespace gcs {
+
+GcsAutoscalerStateManager::GcsAutoscalerStateManager(
+    const ClusterResourceManager &cluster_resource_manager,
+    const GcsResourceManager &gcs_resource_manager,
+    const GcsNodeManager &gcs_node_manager)
+    : cluster_resource_manager_(cluster_resource_manager),
+      gcs_node_manager_(gcs_node_manager),
+      gcs_resource_manager_(gcs_resource_manager),
+      last_cluster_resource_state_version_(0),
+      last_seen_autoscaler_state_version_(0) {}
+
+void GcsAutoscalerStateManager::HandleGetClusterResourceState(
+    rpc::autoscaler::GetClusterResourceStateRequest request,
+    rpc::autoscaler::GetClusterResourceStateReply *reply,
+    rpc::SendReplyCallback send_reply_callback) {
+  RAY_CHECK(request.last_seen_cluster_resource_state_version() <=
+            last_cluster_resource_state_version_);
```

Comment:
huh this no longer true if the gcs fails over in a HA cluster?

Reply:
Hmm, good point. I wasn't really considering this carefully yet. I think we will probably need to add this versioning thing into a persistent storage?

Or we could have a generation number backed by durable KV and gets bumped each time the GcsAutoscalerManager is initialized.

Reply:
let's add a TODO here to handle HA case. we need store an epoch number in kvstore.